### PR TITLE
fix: avoid unwanted watcher creation and reduce being stuck with udf is restarted

### LINKED ISSUE #1001: reduce pod gets stuck when the udf container is restarted.

### PR DESCRIPTION
**Pull request by yhl25**
fixes #1001

### CODE REVIEW

**Review by vigith — APPROVED** [review]

File: pkg/udf/rpc/grpc_reduce.go [hunk 1]
```diff
@@ -75,111 +74,104 @@ func (u *GRPCBasedReduce) WaitUntilReady(ctx context.Context) error {
...
+	)
 
 	// pass key and window information inside the context
 	mdMap := map[string]string{
 		shared.WinStartTime: strconv.FormatInt(partitionID.Start.UnixMilli(), 10),
 		shared.WinEndTime:   strconv.FormatInt(partitionID.End.UnixMilli(), 10),
 	}
 
-	ctx = metadata.NewOutgoingContext(ctx, metadata.New(mdMap))
+	cctx, cancel := context.WithCancel(ctx)
```
Comment (vigith) [inline pkg/udf/rpc/grpc_reduce.go:91]: why do you need a new context?

File: pkg/udf/rpc/grpc_reduce.go [hunk 2]
```diff
@@ -75,111 +74,104 @@ func (u *GRPCBasedReduce) WaitUntilReady(ctx context.Context) error {
...
+		close(responseCh)
 	}()
 
-readLoop:
-	for {
-		select {
-		case msg, ok := <-messageStream:
-			if msg != nil {
+	// create datum from isbMessage and send it to datumCh channel for reduceFn
+	go func() {
```
Comment (whynowy) [inline pkg/udf/rpc/grpc_reduce.go:112]:
```golang
go func() {
  defer close(datumCh)
  for {
    select {
      if xxx {
        return
      }
      xxxx
    }
  }
}()
```

File: pkg/udf/rpc/grpc_reduce.go (hunk 1 above)
Comment (yhl25) [reply pkg/udf/rpc/grpc_reduce.go:91]: to exit from readloop go routine when we return from ApplyReduce.

File: pkg/udf/rpc/grpc_reduce.go (hunk 2 above)
Comment (yhl25) [reply pkg/udf/rpc/grpc_reduce.go:112]: done

File: pkg/udf/rpc/grpc_reduce.go [hunk 3]
```diff
@@ -75,111 +74,103 @@ func (u *GRPCBasedReduce) WaitUntilReady(ctx context.Context) error {
...
+			select {
+			case msg, ok := <-messageStream:
+				// if the messageStream is closed, break the loop
+				if !ok {
+					return
+				}
+				// if the message is nil, break the loop
+				if msg == nil {
+					return
+				}
```
Comment (vigith) [inline pkg/udf/rpc/grpc_reduce.go:125]: reverted this change?

Comment (yhl25) [reply pkg/udf/rpc/grpc_reduce.go]: fixed

File: pkg/udf/rpc/grpc_reduce.go (hunk 1 above)
Comment (whynowy) [reply pkg/udf/rpc/grpc_reduce.go:91]: It should be okay to just do `ctx, cancel := context.WithCancel(ctx)`, do not need to give a new `cctx`.

This is a common patten to make sure everyone in this function watches this `ctx` will properly exit if the function returns, no matter if it's expectedly or unexpectedly.

**Review by whynowy — APPROVED** [review]

**Pull request merged**